types/eslint-scope/index.d.cts

```diff
@@ -0,0 +1,539 @@
+     */
+    writeExpr: ESTree.Expression | null;
+
+    /**
+     * Whether this is a partial reference.
+     */
+    partial: boolean;
+
+    /**
+     * Whether this is an initialization reference.
+     */
+    init: boolean;
+
+    /**
+     * Whether this reference is only read.
+     * @returns True if the reference is read-only.
+     */
+    isReadOnly(): boolean;
+
+    /**
+     * Whether this reference is only written.
+     * @returns True if the reference is write-only.
+     */
+    isWriteOnly(): boolean;
+
+    /**
+     * Whether this reference is read-write.
+     * @returns True if the reference is read-write.
+     */
+    isReadWrite(): boolean;
+}
+
+/**
+ * Represents a variable definition.
+ * @todo extends eslint.Scope.Definition for this class
+ */
+declare class Definition {
+    /**
+     * Creates a new Definition instance.
+     * @param type The type of definition (e.g., 'Variable', 'Parameter').
+     * @param name The identifier node of the definition.
+     * @param node The AST node where the definition occurs.
+     * @param parent The parent node, if applicable.
+     * @param index The index of the definition in a pattern, if applicable.
+     * @param kind The kind of variable (e.g., 'var', 'let', 'const'), if applicable.
+     */
+    constructor(
+        type: eslint.Scope.Definition["type"],
+        name: eslint.Scope.Definition["name"],
+        node: eslint.Scope.Definition["node"],
+        parent: eslint.Scope.Definition["parent"],
+        index: number | null,
+        kind: string | null,
+    );
+
+    /**
+     * The type of definition (e.g., 'Variable', 'Parameter').
+     */
+    type: eslint.Scope.Definition["type"];
+
+    /**
+     * The identifier node of the definition.
+     */
+    name: eslint.Scope.Definition["name"];
+
+    /**
+     * The AST node where the definition occurs.
+     */
+    node: eslint.Scope.Definition["node"];
+
+    /**
+     * The parent node, if applicable.
+     */
+    parent: eslint.Scope.Definition["parent"];
+
+    /**
+     * The index of the definition in a pattern, if applicable.
+     */
+    index: number | null;
+
+    /**
+     * The kind of variable (e.g., 'var', 'let', 'const'), if applicable.
+     */
+    kind: string | null;
+}
+
+/**
+ * Analyzes the scope of an AST.
+ * @param ast The ESTree-compliant AST to analyze.
+ * @param options Options for scope analysis.
+ * @returns The scope manager for the analyzed AST.
+ */
+declare function analyze(ast: ESTree.Program, options?: EslintScope.AnalyzeOptions): ScopeManager;
+
+declare const EslintScope: {
+    analyze: typeof analyze;
+    BlockScope: typeof BlockScope;
+    Definition: typeof Definition;
+    FunctionScope: typeof FunctionScope;
+    GlobalScope: typeof GlobalScope;
+    ModuleScope: typeof ModuleScope;
+    Reference: typeof Reference;
+    Scope: typeof Scope;
+    ScopeManager: typeof ScopeManager;
+    Variable: typeof Variable;
```

Comment:
You don't have to add this, but you may be missing an export of `PatternVisitor`: https://github.com/eslint/js/blob/main/packages/eslint-scope/lib/index.js#L165

Reply:
I think I missed it. I opened another PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/72705 to add `esrecurse`, after that I can rebase this PR and add `PatternVisitor` to `eslint-scope`.